src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/common/util/HttpClientUtil.java

```diff
@@ -285,4 +265,50 @@ public static HttpURLConnection openConnection(URL url)
         }
         return conn;
     }
+
+    private static abstract class BaseRedirectHandler {
+        static final int MAX_REDIRECT_TIMES = 20;
+
+        /**
+         * init an opened connection
+         * @param conn an opened connection
+         */
+        abstract void init(HttpURLConnection conn);
+
+        private HttpURLConnection open(String url, String[] headers)
+            throws IOException, NoSuchAlgorithmException, KeyManagementException {
+            HttpURLConnection conn = openConnection(new URL(url));
+            fillHeaders(conn, headers);
+            init(conn);
+            return conn;
+        }
+
+        HttpURLConnection connect(String url, String[] headers)
+            throws IOException, KeyManagementException, NoSuchAlgorithmException, UrlNotReachableException {
+            HttpURLConnection conn = null;
+            try {
+                for (int i = 0; i < MAX_REDIRECT_TIMES; i++) {
+                    conn = open(url, headers);
+                    conn.connect();
+                    int code = conn.getResponseCode();
+                    if (REDIRECTED_CODE.contains(code)) {
+                        headers = null;
+                        url = conn.getHeaderField("Location");
+                        if (StringUtils.isNotBlank(url)) {
+                            closeConn(conn);
+                            conn = null;
+                            continue;
+                        }
+                        throw new UrlNotReachableException(code + ": redirect but no location");
+                    }
+                    return conn;
+                }
+            } catch (Exception e) {
```

Comment:
move the try catch to for loop?

Reply:
This function doesn't attempt retrying, it will retry 3 times outside.